### ISSUE TITLE
SAML assertion not valid

### ISSUE THREAD

**Issue by blade2005** [jmhale/okta-awscli#36]
I'm getting an error back from the application. 

The error is as stated above. The gist of it is that I'm not seeing SAMLResponse in the html. I have 2 input tags, _xsrfToken and fromURI.

Any reason why I wouldn't have gotten a SAMLResponse back?

**Comment by alex-valchuk** [comment]
I've encountered with the same issue. It's because a page does not contain staticly defined username and password input tags. Instead the source code has javascript that adds those tags dynamicly while loading in a browser.

So the issue is still open

**Comment by jmhale** [comment]
We're not relying on the input tags for username or password to perform authentication. Primary auth and MFA are all done via the API. The only thing we're looking at the HTML for is the presence or absence of the `SAMLReponse` tag, and then pulling the value of the tag, should it exist.

The lack of a `SAMLResponse` tag in the HTML response is likely due to a failed primary or MFA authentication.

Without seeing the full response, I can't be sure, however.

Either way, all of these conditions should be handled individually and provide more meaningful feedback to the user. I'll attempt to reproduce and see if I can find out why the generic lack of SAMLResponse error is being thrown, instead of something more specific.

All that said, this also may have been a transient event within Okta itself.

**Comment by infa-ppatel** [comment]
I was debugging this issue and populated html response from get_saml_assertion().

HTML response points to an error as below:
<div class="unsupported-browser-banner-wrap">
      <div class="unsupported-browser-banner-inner">
        <span class="icon icon-16 icon-only warning-16-yellow"></span>You are using an unsupported browser. For the best experience, update to <a href="https://support.okta.com/help/articles/Knowledge_Article/24532952-Platforms---Browser-and-OS-Support">a supported browser</a>.</div>
    </div>

**Comment by jmhale** [comment]
@infa-ppatel Are you seeing the unsupported browser in the HTML response in conjunction with the lack of a `SAMLResponse` tag?

They definitely sound related. I'm still unable to get a empty `SAMLResponse`, but I also don't see the unsupported browser error in the HTML output of `get_saml_assertion()`

**Comment by infa-ppatel** [comment]
There is no SAMLResponse in HTML response page.
It seems that when Unsupported browser error is encountered it redirects itself to sign-in page and thus no SAMLResponse.

**Comment by abhisheksoniDX** [comment]
@jmhale I am too getting the same error. It was working fine and then suddenly started throwin above error. 
Since you want to see full HTML response:
```
okta-awscli --profile okta-dev-view sts get-caller-identity --okta-profile dev-view
 Enter password: 
 Enter MFA token: 352087
 <!DOCTYPE html>

<!--[if IE 7]><html class="lt-ie10 lt-ie9 lt-ie8"><![endif]-->
<!--[if IE 8]><html class="lt-ie10 lt-ie9"> <![endif]-->
<!--[if IE 9]><html class="lt-ie10"><![endif]-->
<!--[if gt IE 9]><html><![endif]-->
<!--[if !IE]><!--><html><!--<![endif]-->
<head>
<script>if (typeof module === 'object') {window.module = module; module = undefined;}</script>
<title>DataXu - Sign In</title>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport">
<meta content="none" name="robots"/>
<link href="https://ok4static.oktacdn.com/assets/loginpage/css/okta-login-page.min.c2335d687406691ab0663072de302c86.css" rel="stylesheet" type="text/css"/><script>
        var okta = {
            locale: 'en',
            deployEnv: 'PROD'
        };
    </script>
<script>window.okta || (window.okta = {}); okta.cdnUrlHostname = "//ok4static.oktacdn.com"; okta.cdnPerformCheck = false;</script><script>window.okta || (window.okta = {});window.okta.mixpanel = true;window.okta.mixpanelTrackingSamplingFactors = {"_DEFAULT":1.0};</script><script>if (window.module) module = window.module;</script>
</meta></head>
<body class="auth okta-container">
<!--[if gte IE 8]>
  <![if lte IE 9]>

    <style>
    .unsupported-browser-banner-wrap {
      padding: 20px;
      border: 1px solid #ddd;
      background-color: #f3fbff;
    }
    .unsupported-browser-banner-inner {
      position: relative;
      width: 735px;
      margin: 0 auto;
      text-align: left;
    }
    .unsupported-browser-banner-inner .icon {
      vertical-align: top;
      margin-right: 20px;
      display: inline-block;
      position: static !important;
    }
    .unsupported-browser-banner-inner a {
      text-decoration: underline;
    }
    </style>

    <div class="unsupported-browser-banner-wrap">
      <div class="unsupported-browser-banner-inner">
        <span class="icon icon-16 icon-only warning-16-yellow"></span>You are using an unsupported browser. For the best experience, update to <a href="https://support.okta.com/help/articles/Knowledge_Article/24532952-Platforms---Browser-and-OS-Support">a supported browser</a>.</div>
    </div>

  <![endif]>
<![endif]-->
<!--[if IE 8]> <div id="login-bg-image-ie8" class="login-bg-image" data-se="login-bg-image"></div> <![endif]-->
<!--[if (gt IE 8)|!(IE)]><!--> <div class="login-bg-image" data-se="login-bg-image" id="login-bg-image"></div> <!--<![endif]-->
<!-- hidden form for reposting fromURI for X509 auth -->
<form action="/login/cert" id="x509_login" method="post" name="x509_login" style="display:none;">
<input class="hide" name="_xsrfToken" type="hidden" value="null"/><input class="hidden" id="fromURI" name="fromURI" type="hidden" value="/policy/second-factor?fromURI=%2Fapp%2Famazon_aws%2Fexk9mqjao1NEIbACB1t7%2Fsso%2Fsaml&amp;ruleId=plr9v96gpw5i0aJ9l1t7&amp;appInstanceId=0oa9mqjao2D2uSudU1t7"/>
</form>
<div class="content">
<style type="text/css">
    .noscript-msg {
        background-color: #fff;
        border-color: #ddd #ddd #d8d8d8;
        box-shadow:0 2px 0 rgba(175, 175, 175, 0.12);
        text-align: center;
        width: 398px;
        min-width: 300px;
        margin: 200px auto;
        border-radius: 3px;
        border-width: 1px;
        border-style: solid;
    }

    .noscript-content {
        padding: 42px;
    }

    .noscript-content h2 {
        padding-bottom: 20px;
    }

    .noscript-content h1 {
        padding-bottom: 25px;
    }

    .noscript-content a {
        background: transparent;
        box-shadow: none;
        display: table-cell;
        vertical-align: middle;
        width: 314px;
        height: 50px;
        line-height: 36px;
        color: #fff;
        background: linear-gradient(#007dc1, #0073b2), #007dc1;
        border: 1px solid;
        border-color: #004b75;
        border-bottom-color: #00456a;
        box-shadow: rgba(0, 0, 0, 0.15) 0 1px 0, rgba(255, 255, 255, 0.1) 0 1px 0 0 inset;
        -webkit-border-radius: 3px;
        border-radius: 3px;
    }

    .noscript-content a:hover {
        background: #007dc1;
        cursor: hand;
        text-decoration: none;
    }
</style>
<noscript>
<div class="noscript-msg" id="noscript-msg">
<div class="noscript-content">
<h2>Javascript is required</h2>
<h1>Javascript is disabled on your browser.&amp;nbspPlease enable Javascript and refresh this page.</h1>
<a href=".">Refresh</a>
</div>
</div>
</noscript>
<div id="signin-container"></div>
<div class="auth-container main-container" id="okta-sign-in" style="display:none">
<div class="unsupported-message" id="unsupported-onedrive" style="display:none">
<h2 class="o-form-head">Your OneDrive version is not supported</h2>
<p>Upgrade now by installing the OneDrive for Business Next Generation Sync Client to login to Okta</p>
<a class="button button-primary" href="https://support.okta.com/help/articles/Knowledge_Article/Upgrading-to-OneDrive-for-Business-Next-Generation-Sync-Client" target="_blank">
          Learn how to upgrade</a>
</div>
<div class="unsupported-message" id="unsupported-cookie" style="display:none">
<h2 class="o-form-head">Cookies are required</h2>
<p>Cookies are disabled on your browser. Please enable Cookies and refresh this page.</p>
<a class="button button-primary" href="." target="_blank">
              Refresh</a>
</div>
</div>
</div>
<div class="footer">
<div class="footer-container clearfix">
<p class="copyright">Powered by <a class="inline-block notranslate" href="http://www.okta.com/">Okta</a></p>
<p class="privacy-policy"><a class="inline-block margin-l-10" href="/privacy" target="_blank">Privacy Policy</a></p>
</div>
</div>
<script type="text/javascript">function runLoginPage (fn) {var mainScript = document.createElement('script');mainScript.src = 'https://ok4static.oktacdn.com/assets/js/mvc/loginpage/initLoginPage.pack.28480ea192eb1871ce16e253fbd87728.js?v=1';document.getElementsByTagName('head')[0].appendChild(mainScript);fn && mainScript.addEventListener('load', function () { setTimeout(fn, 1) });}</script><script type="text/javascript">
(function(){

  var baseUrl = 'https\x3A\x2F\x2Fdataxu.okta.com';
  var suppliedRedirectUri = '';
  var repost = false;
  var stateToken = '';
  var fromUri = '\x2Fpolicy\x2Fsecond\x2Dfactor\x3FfromURI\x3D\x252Fapp\x252Famazon_aws\x252Fexk9mqjao1NEIbACB1t7\x252Fsso\x252Fsaml\x26ruleId\x3Dplr9v96gpw5i0aJ9l1t7\x26appInstanceId\x3D0oa9mqjao2D2uSudU1t7';
  var username = '';
  var rememberMe = true;
  var smsRecovery = false;
  var callRecovery = false;
  var emailRecovery = true;
  var usernameLabel = 'Username';
  var usernameInlineLabel = '';
  var passwordLabel = 'Password';
  var passwordInlineLabel = '';
  var signinLabel = 'Sign\x20In';
  var forgotpasswordLabel = 'Forgot\x20password\x3F';
  var unlockaccountLabel = 'Unlock\x20account\x3F';
  var helpLabel = 'Help';
  var orgSupportPhoneNumber = '857\x2D366\x2D4141';
  var hideSignOutForMFA = false;
  var loginPageUrlRedirect = '';
  var enableUrlFixForEmbeddedBrowsers = false;
  var footerHelpTitle = 'Need\x20help\x20signing\x20in\x3F';
  var recoveryFlowPlaceholder = 'Email\x20or\x20Username';
  var signOutUrl = '';
  var authScheme = 'OAUTH2';
  var securityImage = true;
  var windowsVerify = false;
    windowsVerify = true;
    var selfServiceUnlock = false;
 
  var preventBrowserFromSavingOktaPassword = false;
  var enableMixpanelTracking = false;

  var autoPush = false;
    autoPush = true;
  var publishToAccountChooser = false;
  var accountChooserDiscoveryUrl = null;
 
    publishToAccountChooser = true;
    accountChooserDiscoveryUrl = 'https://login.okta.com/discovery/iframe.html';

  // In case of custom app login, the uri is already absolute, so we must not attach baseUrl
  var redirectUri;
  if (isAbsoluteUri(fromUri)) {
      redirectUri = fromUri;
  } else {
      redirectUri = baseUrl + fromUri;
  }
  var customButtons;
  var customLinks = [];
  var linkParams;
  var idpDiscovery;
  var idpDiscoveryRequestContext;
  var hasPasswordlessPolicy = false;
  var showPasswordToggleOnSignInPage = false;
  var hasOAuth2ConsentFeature = false;
  var consentFunc;
  var hasMfaAttestationFeature = false;
  var registration = false;
  var webauthn = false;
  var loginPageConfig = {
    fromUri: fromUri,
    repost: repost,
    redirectUri: redirectUri,
    isMobileClientLogin: false,
    isMobileSSO: false,

    linkParams: linkParams,
    hasChromeOSFeature: false,
    showLinkToAppStore: false,
    publishToAccountChooser: publishToAccountChooser,
    accountChooserDiscoveryUrl: accountChooserDiscoveryUrl,
    preventBrowserFromSavingOktaPassword: preventBrowserFromSavingOktaPassword,
    enableMixpanelTracking: enableMixpanelTracking,
    enableUrlFixForEmbeddedBrowsers: enableUrlFixForEmbeddedBrowsers,
    loginPageUrlRedirect: loginPageUrlRedirect,
    mfaAttestation: hasMfaAttestationFeature,
    signIn: {
      el: '#signin-container',
      baseUrl: baseUrl,
      logo: 'https://ok4static.oktacdn.com/bc/image/fileStoreRecord?id=fs03lealwhfloJUKr1t7',
      logoText: 'DataXu',
      helpSupportNumber: orgSupportPhoneNumber,
      stateToken: stateToken,
      username: username,
      signOutLink: signOutUrl,
      consent: consentFunc,
      authScheme: authScheme,
      relayState: fromUri,
      idpDiscovery: {
        requestContext: idpDiscoveryRequestContext
      },
      features: {
        router: true,
        securityImage: securityImage,
        rememberMe: rememberMe,
        autoPush: autoPush,
        webauthn: webauthn,
        smsRecovery: smsRecovery,
        callRecovery: callRecovery,
        emailRecovery: emailRecovery,
        windowsVerify: windowsVerify,
        selfServiceUnlock: selfServiceUnlock,
        multiOptionalFactorEnroll: true,
        deviceFingerprinting: true,
        trackTypingPattern: false,
        hideSignOutLinkInMFA: hideSignOutForMFA,
        customExpiredPassword: true,
        idpDiscovery: idpDiscovery,
        passwordlessAuth: hasPasswordlessPolicy,
        consent: hasOAuth2ConsentFeature,
        showPasswordToggleOnSignInPage: showPasswordToggleOnSignInPage,
        registration: registration
      },

      assets: {
        baseUrl: okta.cdnUrlHostname + '/assets/js/mvc/loginpage/i18n'
      },

      language: okta.locale,
      i18n: {},

      customButtons: customButtons,

      helpLinks: {
        help: '',
        forgotPassword: '',
        unlock: '',
        custom: customLinks
      }
    }
  };

  loginPageConfig.signIn.i18n[okta.locale] = {
    
    'primaryauth.username.placeholder': usernameLabel,
    'primaryauth.username.tooltip': usernameInlineLabel,
    'primaryauth.password.placeholder': passwordLabel,
    'primaryauth.password.tooltip': passwordInlineLabel,
    'primaryauth.title': signinLabel,
    'forgotpassword': forgotpasswordLabel,
    'unlockaccount': unlockaccountLabel,
    'help': helpLabel,
    'needhelp': footerHelpTitle,
    'password.forgot.email.or.username.placeholder': recoveryFlowPlaceholder,
    'password.forgot.email.or.username.tooltip': recoveryFlowPlaceholder,
    'account.unlock.email.or.username.placeholder': recoveryFlowPlaceholder,
    'account.unlock.email.or.username.tooltip': recoveryFlowPlaceholder
  };

  function isOldWebBrowserControl() {
    // We no longer support IE7. If we see the MSIE 7.0 browser mode, it's a good signal
    // that we're in a windows embedded browser.
    if (navigator.userAgent.indexOf('MSIE 7.0') === -1) {
      return false;
    }

    // Because the userAgent is the same across embedded browsers, we use feature
    // detection to see if we're running on older versions that do not support updating
    // the documentMode via x-ua-compatible.
    return document.all && !window.atob;
  }

  function isAbsoluteUri(uri) {
    var pat = /^https?:\/\//i;
    return pat.test(uri);
  }

  var unsupportedContainer = document.getElementById('okta-sign-in');

  // Old versions of WebBrowser Controls (specifically, OneDrive) render in IE7 browser
  // mode, with no way to override the documentMode. In this case, inform the user they need
  // to upgrade.
  if (isOldWebBrowserControl()) {
    document.getElementById('unsupported-onedrive').removeAttribute('style');
    unsupportedContainer.removeAttribute('style');
  }
  else if (!navigator.cookieEnabled) {
    document.getElementById('unsupported-cookie').removeAttribute('style');
    unsupportedContainer.removeAttribute('style');
  }
  else {
    unsupportedContainer.parentNode.removeChild(unsupportedContainer);
    runLoginPage(function () {
      OktaLogin.initLoginPage(loginPageConfig);
    });
  }

}());
</script>
<script>
  window.addEventListener('load', function(event) {
    function applyStyle(id, style) {
      if (style) {
        var el = document.getElementById(id);
        if (el) {
          el.setAttribute('style', style);
        }
      }
    }
    applyStyle('login-bg-image', "background-image: url('https://ok4static.oktacdn.com/bc/image/fileStoreRecord?id=fs023fqrdjW1GICl11t7')");
    applyStyle('login-bg-image-ie8', "filter: progid:DXImageTransform.Microsoft.AlphaImageLoader(src='https://ok4static.oktacdn.com/bc/image/fileStoreRecord?id=fs023fqrdjW1GICl11t7', sizingMethod='scale')");
  });
</script>
</body>
</html>
```

**Comment by rtkjbillo** [comment]
Hi, I was experiencing the same issue (where no SAMLResponse `<input>` field was present after selecting the app) and believe there may be a workaround in terms of Okta administration. There could be two things at play here:

1) If the Okta admin has set a requirement for a MFA re-prompt when logging into the AWS app, the page title will contain "Extra Verification", so the same technique from https://github.com/Nike-Inc/gimme-aws-creds/blob/master/gimme_aws_creds/okta.py#L443 should be able to be used to get the stateToken out of the JavaScript on the page, then re-call the `get_saml_response` function.

2) If the Okta admin has set a requirement for a password re-prompt when logging into the AWS app, there will be no `stateToken` value or `SAMLResponse` input element on the webpage, and for all intents and purposes this will look like the "unsupported browser" message (JavaScript later generates the necessary password field, if you request this page in a browser with the appropriate cookie set.) In my case, I was able to remove the password re-prompt from the AWS app in Okta administration, and the response from Okta once again contained a SAMLResponse `<input>` element.

Hope this helps in terms of troubleshooting.

**Comment by jmhale** [comment]
Thanks for that @rtkjbillo. I think that's definitely getting closer to the issue. I'm able to reproduce the `ERROR - SAML assertion not valid:` if I have the app in the Okta config set to re-prompt for MFA. 

To be fair, we do call out per-application MFA as a unsupported feature. That said, I took a look at what `gimme-aws-creds` is doing, and it shouldn't be too difficult to replicate it here.

Oddly enough, I don't get the SAML assertion error when I set the Okta app to re-prompt for password.

@infa-ppatel @blade2005 @abhisheksoniDX @alex-valchuk - Do any of your problematic Okta apps have re-prompt for password or MFA (prompting for either of them *after* the app is selected)?

I won't have time to work on this for the next week or two, so if someone wants to take a crack here, please have at it.

**Comment by infa-ppatel** [comment]
Yes. I can confirm that we use MFA for AWS within okta flow.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by abhisheksoniDX** [comment]
@jmhale  @infa-ppatel  @rtkjbillo Yes, I too do confirm this. I was using extra MFA for dev application and that was causing the issue.

The moment I removed that extra MFA, it worked fine.

I can make the required changes in the tool. Just want confirmation that no one else doing it. Don't want to do duplicate work.

**Comment by jmhale** [comment]
This should be fixed as of v0.4.0 (#107 / #108 ). Closing
